### PR TITLE
consensus/pbft: verify preprepare message

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: consensus/pbft/backends/simple/backend.go [hunk 1]
```diff
@@ -181,9 +182,16 @@ func (sb *simpleBackend) EventMux() *event.TypeMux {
...
 // Verify implements pbft.Backend.Verify
-func (sb *simpleBackend) Verify(proposal *pbft.Proposal) (bool, error) {
-	// not implemented
-	return true, nil
+func (sb *simpleBackend) Verify(proposal *pbft.Proposal) error {
+	// decode the proposal to block
+	block := &types.Block{}
+	err := rlp.DecodeBytes(proposal.Payload, block)
+	if err != nil {
+		log.Warn("decode block error", "err", err)
```
Comment (tailingchen) [inline consensus/pbft/backends/simple/backend.go]: use sb.logger

File: consensus/pbft/core/preprepare.go [hunk 2]
```diff
@@ -46,6 +46,11 @@ func (c *core) handlePreprepare(preprepare *pbft.Preprepare, src pbft.Validator)
 		return errFutureMessage
 	}
 
+	if err := c.backend.Verify(preprepare.Proposal); err != nil {
```
Comment (bailantaotao) [inline consensus/pbft/core/preprepare.go]: 要不要先檢查proposer再做verify啊？

verify感覺比較耗時xd

File: consensus/pbft/backends/simple/backend.go [hunk 3]
```diff
@@ -143,13 +145,16 @@ func (sb *simpleBackend) Commit(proposal *pbft.Proposal) error {
...
 	if sb.commit != nil {
-		go func() {
-			sb.commit <- block.Hash()
-		}()
+		sb.commitErr = make(chan error, 1)
+		closeCommitErr := func() {
+			close(sb.commitErr)
+		}
+		defer closeCommitErr()
+		sb.commit <- block.Hash()
```
Comment (tailingchen) [inline consensus/pbft/backends/simple/backend.go]: add comment

File: consensus/pbft/core/preprepare.go (hunk 2 above)
Comment (markya0616) [inline consensus/pbft/core/preprepare.go]: OK

File: consensus/pbft/backends/simple/backend.go (hunk 1 above)
Comment (markya0616) [inline consensus/pbft/backends/simple/backend.go]: OK

File: consensus/pbft/backends/simple/backend.go (hunk 3 above)
Comment (markya0616) [inline consensus/pbft/backends/simple/backend.go]: ok

File: consensus/pbft/backends/simple/engine.go [hunk 4]
```diff
@@ -230,8 +237,9 @@ func (sb *simpleBackend) Prepare(chain consensus.ChainReader, header *types.Head
 		return consensus.ErrUnknownAncestor
 	}
 	// Ensure the extra data has all it's components
-	length := len(header.Extra)
-	header.Extra = header.Extra[:length-extraSeal]
+	length := len(parent.Extra)
+	header.Extra = make([]byte, 0)
```
Comment (alanchchen) [inline consensus/pbft/backends/simple/engine.go]: `append` can handle nil value. This could be removed.

Comment (alanchchen) [inline consensus/pbft/backends/simple/engine.go]: `append` can handle `nil` value. This could be removed.

Comment (markya0616) [inline consensus/pbft/backends/simple/engine.go]: Just in case `header.Extra` is not empty, so we reset it first and assign the correct values.

**Pull request merged**